### PR TITLE
pkp/pkp-lib#6282 Adding warnings if someone try to disable every forms and submissions language

### PR DESCRIPTION
**Pull request by henriqueramos**
# Evidence

https://user-images.githubusercontent.com/1243634/106774664-1d82d700-6621-11eb-9da3-52a31ad03d5f.mp4

### CODE REVIEW

File: locale/en_US/manager.po [hunk 1]
```diff
@@ -1138,6 +1138,12 @@ msgstr "{$locale} locale reloaded for {$contextName}."
 msgid "notification.localeSettingsSaved"
 msgstr "Locale settings saved."
 
+msgid "notification.localeSettingsCannotBeSaved.emptyFormsOption"
+msgstr "Locale settings cannot be saved. Select at least one language at Forms"
```
Comment (NateWr) [inline locale/en_US/manager.po:1142]: Change this to: "The language setting could not be saved. At least one language must be enabled for each option."

Then you can reuse that locale key for both cases.

**Review by NateWr — CHANGES_REQUESTED** [review]
Looks great, thanks @henriqueramos! I have one more request that has just occurred to me. It's also possible to modify the `supportedFormLocales` and `supportedSubmissionLocales` setting through the API.

To catch these cases, can you add a validation check to make sure that at least one locale is selected for these properties? THat could be added here:

https://github.com/pkp/pkp-lib/blob/master/classes/services/PKPContextService.inc.php#L301

Hopefully you will recognize the Laravel validator, but let me know if you have any questions about adding these validation checks.

**Comment by henriqueramos** [comment]
> Looks great, thanks @henriqueramos! I have one more request that has just occurred to me. It's also possible to modify the `supportedFormLocales` and `supportedSubmissionLocales` setting through the API.
> 
> To catch these cases, can you add a validation check to make sure that at least one locale is selected for these properties? THat could be added here:
> 
> https://github.com/pkp/pkp-lib/blob/master/classes/services/PKPContextService.inc.php#L301
> 
> Hopefully you will recognize the Laravel validator, but let me know if you have any questions about adding these validation checks.

`Illuminate\Validation\Validator` rules already take care of non-valid `supportedFormLocales` and `supportedSubmissionLocales` values.

Screenshot from 2021-02-08 07-51-44

**Review by NateWr — CHANGES_REQUESTED** [review]
I was still able to make a PUT request to the REST API that added an unsupported locale. Here are the reproduction steps:

1. Go to Administration > Site Settings > Language.
2. Add support for en_US and pt_BR.
3. Make a `PUT` request to `/contexts/1` with the following payload:

```
{
  supportedLocales: ['en_US'],
  supportedFormLocales: ['pt_BR'],
}
```

This should throw a validation error because a `supportedFormLocale` should not include a locale that isn't in `supportedFormLocales`.

File: classes/services/PKPContextService.inc.php [hunk 2]
```diff
@@ -299,6 +299,37 @@ public function validate($action, $props, $allowedLocales, $primaryLocale) {
...
+			};
+
+			$supportedLocales = $getCurrentPropValue('supportedLocales');
+			$supportedFormLocales = $getCurrentPropValue('supportedFormLocales');
+			$supportedSubmissionLocales = $getCurrentPropValue('supportedSubmissionLocales');
+
+			$unsupportedFormLocales = array_diff($supportedFormLocales, $supportedLocales);
+			$unsupportedSubmissionLocales = array_diff($supportedSubmissionLocales, $supportedLocales);
+
+			if ($unsupportedFormLocales !== [] && !$validator->errors()->has('supportedFormLocales')) {
```
Comment (NateWr) [inline classes/services/PKPContextService.inc.php:324]: I think that we need to only throw these validation errors if a prop is passed. So only run these checks if `isset($props['supportedFormLocales'])`. That will allow a situation where the `supportedLocales` are changed and we automatically remove the `supportedFormLocales`.

The same is true for `supportedSubmissionLocales` below.

**Review by NateWr — CHANGES_REQUESTED** [review]
Great!  Just a couple of minor issues to sort out and I think we're there!

File: classes/services/PKPContextService.inc.php [hunk 3]
```diff
@@ -299,6 +299,43 @@ public function validate($action, $props, $allowedLocales, $primaryLocale) {
...
 
+		// Ensure that the form and submission locales are available for the context
+		$validator->after(function($validator) use ($action, $props) {
+			$context = Application::get()->getRequest()->getContext();
+			if (isset($props['id']) && $context->getId() !== $props['id']) {
+				$context = $this->get($props['id']);
+			}
+
+			$getCurrentPropValue = function($prop) use ($props, $context) {
+				if (isset($props[$prop])) {
```
Comment (asmecher) [inline classes/services/PKPContextService.inc.php:310]: The `??` operator would make this function a lot more concise:
```
$props[$prop] ?? $context->getData($prop)
```

File: classes/services/PKPContextService.inc.php [hunk 4]
```diff
@@ -299,6 +299,43 @@ public function validate($action, $props, $allowedLocales, $primaryLocale) {
...
+
+				return $context->getData($prop);
+			};
+
+			$supportedLocales = $getCurrentPropValue('supportedLocales');
+			$supportedFormLocales = $getCurrentPropValue('supportedFormLocales');
+			$supportedSubmissionLocales = $getCurrentPropValue('supportedSubmissionLocales');
+
+			$unsupportedFormLocales = array_diff($supportedFormLocales, $supportedLocales);
+			$unsupportedSubmissionLocales = array_diff($supportedSubmissionLocales, $supportedLocales);
```
Comment (asmecher) [inline classes/services/PKPContextService.inc.php:322]: For readability, I'd suggest moving this down a little closer to where it's used.

File: classes/services/PKPContextService.inc.php [hunk 5]
```diff
@@ -299,6 +299,43 @@ public function validate($action, $props, $allowedLocales, $primaryLocale) {
...
+
+			if (isset($props['supportedFormLocales'])
```
Comment (asmecher) [inline classes/services/PKPContextService.inc.php:324]: Each of these checks could use a short comment to explain what we're looking for, e.g.:
```
// If any of the selected form locales is not in the supported list, add an error.
```

File: controllers/grid/languages/LanguageGridHandler.inc.php [hunk 6]
```diff
@@ -103,8 +103,22 @@ function saveLanguageSetting($args, $request) {
 						$key = array_search($locale, $supportedSubmissionLocales);
 						if ($key !== false) unset($supportedSubmissionLocales[$key]);
 						$supportedSubmissionLocales = array_values($supportedSubmissionLocales);
+						if ($supportedSubmissionLocales == []) {
+							return new JSONMessage(false, __('notification.localeSettingsCannotBeSaved'));
+						}
 						$context = $contextService->edit($context, ['supportedSubmissionLocales' => $supportedSubmissionLocales], $request);
 					}
+
+					if ($settingName == 'supportedSubmissionLocales') {
```
Comment (asmecher) [inline controllers/grid/languages/LanguageGridHandler.inc.php:112]: I think this check is added too late -- after the context is updated.

**Review by asmecher — CHANGES_REQUESTED** [review]
Just a few comments -- thanks, Henrique!

File: classes/services/PKPContextService.inc.php (hunk 3 above)
Comment (henriqueramos) [reply classes/services/PKPContextService.inc.php:310]: Indeed, but I ain't the biggest fan of the [null coalescing operator](https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op). So I don't use it a lot, always opting for `if` usage instead; letting the code more "readable" and verbose.

File: controllers/grid/languages/LanguageGridHandler.inc.php (hunk 6 above)
Comment (henriqueramos) [reply controllers/grid/languages/LanguageGridHandler.inc.php:112]: We've made some test scenarios where tried to disable all submissions checkboxes, everyone received the locale settings cannot be saved notification.

@NateWr we could do a double check here and see if there's any other scenario to test, what do you think?

File: classes/services/PKPContextService.inc.php (hunk 5 above)
Comment (henriqueramos) [reply classes/services/PKPContextService.inc.php:327]: Added it before the `if`. ;)

File: controllers/grid/languages/LanguageGridHandler.inc.php (hunk 6 above)
Comment (NateWr) [reply controllers/grid/languages/LanguageGridHandler.inc.php:112]: I think the use-case that Alec has identified here is when the supported submission locales are disabled automatically as part of the form locales: https://github.com/pkp/pkp-lib/blob/f704ad0a67bbf12f79cdd67db2509ed25bd73c7d/controllers/grid/languages/LanguageGridHandler.inc.php#L100-L110

So an issue might arise in the following case:

1. en_US enabled for all locales
2. fr_CA enabled for UI + Form but not submission
3. Disable Form locales for en_US

In such a case, the journal may end up with no supported submission locales. Can you test that out?

File: classes/services/PKPContextService.inc.php (hunk 3 above)
Comment (asmecher) [reply classes/services/PKPContextService.inc.php:315]: We have room for that kind of stylistic preference in such a large codebase :)

**Pull request closed without merging**